pkg/asset/machines/alibabacloud/machinesets.go

```diff
@@ -0,0 +1,84 @@
+package alibabacloud
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/openshift/installer/pkg/types"
+	"github.com/openshift/installer/pkg/types/alibabacloud"
+	machineapi "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// MachineSets returns a list of machinesets for a machinepool.
+func MachineSets(clusterID string, config *types.InstallConfig, pool *types.MachinePool, role, userDataSecret string) ([]*machineapi.MachineSet, error) {
+	if configPlatform := config.Platform.Name(); configPlatform != alibabacloud.Name {
+		return nil, fmt.Errorf("non-AlibabaCloud configuration: %q", configPlatform)
+	}
+	if poolPlatform := pool.Platform.Name(); poolPlatform != alibabacloud.Name {
+		return nil, fmt.Errorf("non-AlibabaCloud machine-pool: %q", poolPlatform)
+	}
+	platform := config.Platform.AlibabaCloud
+	mpool := pool.Platform.AlibabaCloud
+	azs := mpool.Zones
+
+	total := int64(0)
+	if pool.Replicas != nil {
+		total = *pool.Replicas
+	}
+	numOfAZs := int64(len(azs))
+	var machinesets []*machineapi.MachineSet
+	for idx, az := range azs {
+		replicas := int32(total / numOfAZs)
+		if int64(idx) < total%numOfAZs {
+			replicas++
+		}
+
+		provider, err := provider(clusterID, platform, mpool, idx, role, userDataSecret)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create provider")
+		}
+		name := fmt.Sprintf("%s-%s-%s", clusterID, pool.Name, strings.TrimPrefix(az, fmt.Sprintf("%s-", platform.Region)))
+		mset := &machineapi.MachineSet{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "machine.openshift.io/v1beta1",
+				Kind:       "MachineSet",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "openshift-machine-api",
+				Name:      name,
+				Labels: map[string]string{
+					"machine.openshift.io/cluster-api-cluster": clusterID,
```

Comment:
I think this string must match the version from https://github.com/AliyunContainerService/cluster-api-provider-alibabacloud/blob/master/pkg/apis/alibabacloudprovider/v1beta1/register.go#L38

I think this needs to be `alibabacloudmachineproviderconfig.openshift.io`